### ISSUE TITLE
Implementing iterators in PyMappingProtocol

### ISSUE THREAD

**Issue by sebpuetz** [PyO3/pyo3#611]
## 🐛 Bug Reports

When reporting a bug, please provide the following information. If this is not a bug report you can just discard this template.

### 🌍 Environment

 - Your operating system and version: Ubuntu 18.04
 - Your python version: 3.73
 - How did you install python (e.g. apt or pyenv)? Did you use a virtualenv?: pyenv
 - Your rust version (`rustc --version`): rustc 1.38.0-nightly (311376d30 2019-07-18)
 - Are you using the latest pyo3 version? Have you tried using latest master (replace `version = "0.x.y"` with `git = "https://github.com/PyO3/pyo3")?` yes

### 💥 Reproducing

Implementing `PyIterProtocol` for a type implementing `PyMappingProtocol` seems to overwrite `__iter__` in Python. Invoking an iterator through any kind of statement calls `PyIterProtocol::__iter__`. `PyMappingProtocol::__iter__` is not accessible. Checking whether an element is in `mapping` through the `in` keyword runs an iterator and compares items instead of using `__contains__`. `__contains__` can be invoked explicitly through `PyMapping.__contains__`. 

If the implementation of `PyIterProtocol` is removed for `PyMapping`, statements such as:

~~~Python
import pymapping
mapping = pymapping.Mapping(["a", "b", "c", "d"]) 
for k in mapping:
    print(k)

print("a" in mapping)
~~~

fail by stating that `mapping` is not iteratable. Although, the iterator can be called explicitly through `PyMapping.__iter__` but using `iter()` on a `PyMapping` object fails:

~~~Python
In [1]: import pymapping                                                                                                                                                                             

In [2]: mapping = pymapping.Mapping(["a", "b", "c", "d"])                                                                                                                                            

In [3]: iter(mapping)                                                                                                                                                                                
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-73d010d9f2e9> in <module>
----> 1 iter(mapping)

TypeError: 'Mapping' object is not iterable

In [4]: mapping.__iter__()                                                                                                                                                                           
own_iter
Out[4]: <MappingIter at 0x7f234178de40>
~~~

~~~Rust
#![feature(specialization)]

#[pymodule]
fn pymapping(_py: Python, m: &PyModule) -> PyResult<()> {
    m.add_class::<PyMapping>()?;
    Ok(())
}

use pyo3::prelude::*;
use pyo3::{PyIterProtocol, PyMappingProtocol};

use std::collections::HashMap;
use std::vec;

#[pyclass(name = Mapping)]
pub struct PyMapping {
    map: HashMap<String, usize>,
}

#[pymethods]
impl PyMapping {
    #[new]
    #[args(mmap = false)]
    fn __new__(obj: &PyRawObject, keys: Vec<String>) -> PyResult<()> {
        let map = keys.into_iter().enumerate().map(|(x, y)| (y, x)).collect();

        obj.init(PyMapping { map });

        Ok(())
    }
}

#[pyproto]
impl PyMappingProtocol for PyMapping {
    fn __len__(&self) -> PyResult<usize> {
        Ok(self.map.len())
    }

    fn __getitem__(&self, key: String) -> PyResult<usize> {
        Ok(self.map.get(&key).cloned().unwrap_or_default())
    }

    fn __iter__(&self, py: Python) -> PyResult<PyObject> {
        println!("mapping iter");
        let iter = IntoPy::into_py(
            Py::new(py, PyMappingIter::new(self.map.iter().map(|(k, _)| k.to_owned()).collect()))?,
            py,
        );

        Ok(iter)
    }

    fn __contains__(&self, word: String) -> PyResult<bool> {
        println!("contains");
        Ok(self.map.contains_key(&word))
    }
}

#[pyproto]
impl PyIterProtocol for PyMapping {
    fn __iter__(slf: PyRefMut<Self>) -> PyResult<PyObject> {
        let mapping = &*slf;
        println!("iter protocol");
        let gil = Python::acquire_gil();
        let py = gil.python();
        let iter = IntoPy::into_py(
            Py::new(py, PyMappingIter::new(mapping.map.iter().map(|(k, _)| k.to_owned()).collect()))?,
            py,
        );

        Ok(iter)
    }
}

#[pyclass(name = MappingIter)]
pub struct PyMappingIter {
    v: vec::IntoIter<String>,
}

impl PyMappingIter {
    pub fn new(v: Vec<String>) -> Self {
        PyMappingIter { v: v.into_iter() }
    }
}

#[pyproto]
impl PyIterProtocol for PyMappingIter {
    fn __iter__(slf: PyRefMut<Self>) -> PyResult<Py<PyMappingIter>> {
        Ok(slf.into())
    }

    fn __next__(mut slf: PyRefMut<Self>) -> PyResult<Option<String>> {
        let slf = &mut *slf;
        Ok(slf.v.next())
    }
}
~~~

Please also write what exact flags are required to reproduce your results.

**Comment by kngwyu** [comment]
Thank you for reporting.

**Comment by Alexander-N** [comment]
I guess the problem is that `tp_iter` has to be implemented for any iterable type, but is only present in [PyIterIterProtocol](https://github.com/PyO3/pyo3/blob/17963356/src/class/iter.rs#L78-L91).

**Comment by sebpuetz** [comment]
Would that explain `__contains__` being ignored tho? As far as I can tell, it's only accessible on `PyMapping` types through explicit calls to `__contains__()`.

**Comment by Alexander-N** [comment]
Thanks, your analysis looks spot-on to me! I guess we should remove `__iter__` and `__contains__` from `PyMappingProtocol`.

**Comment by Alexander-N** [comment]
Closed by #644.
